packages/core/contracts/oracle/implementation/StakerSnapshot.sol

```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "./VotingToken.sol";
+import "../../common/implementation/Testable.sol";
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Arrays.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+
+import "hardhat/console.sol";
+
+contract StakerSnapshot is Ownable, Testable {
+    using Arrays for uint256[];
+    using Counters for Counters.Counter;
+
+    /****************************************
+     *           STAKING TRACKERS           *
+     ****************************************/
+
+    uint256 public emissionRate;
+    uint256 public cumulativeStaked;
+    uint256 public rewardPerTokenStored;
+    uint256 public lastUpdateTime;
+
+    uint256 unstakeCoolDown;
+
+    struct VoterStake {
+        uint256 cumulativeStaked;
+        uint256 rewardsPaidPerToken;
+        uint256 outstandingRewards;
+        uint256 unstakeTime;
+        uint256 requestUnstake;
+        uint256 lastRequestIndexConsidered;
+        Snapshots snapshots;
+    }
+
+    mapping(address => VoterStake) public voterStakes;
+
+    struct Snapshots {
+        uint256[] ids;
+        uint256[] values;
+    }
+
+    Snapshots private _cumulativeStakedSnapshots;
+
+    // Snapshot ids increase monotonically, with the first value being 1. An id of 0 is invalid.
+    Counters.Counter private _currentSnapshotId;
+
+    // Reference to the voting token.
+    VotingToken public votingToken;
+
+    event Snapshot(uint256 id);
+
+    constructor(
+        uint256 _emissionRate,
+        uint256 _unstakeCoolDown,
+        address _votingToken,
+        address _timerAddress
+    ) Testable(_timerAddress) {
+        emissionRate = _emissionRate;
+        unstakeCoolDown = _unstakeCoolDown;
+        votingToken = VotingToken(_votingToken);
+    }
+
+    // Pulls tokens from users wallet and stakes them.
+    function stake(uint256 amount) public {
+        _updateTrackers(msg.sender);
+        updateSnapshot(msg.sender);
+
+        voterStakes[msg.sender].cumulativeStaked += amount;
+        cumulativeStaked += amount;
+
+        votingToken.transferFrom(msg.sender, address(this), amount);
+    }
+
+    function updateSnapshot(address voterAddress) public {
+        _updateAccountSnapshot(voterAddress);
+        _updateTotalStakedSnapshot();
+    }
+
+    function requestUnstake(uint256 amount) public {
+        _updateTrackers(msg.sender);
+        // Staker signals that they want to unstake. After signalling, their total voting balance is decreased by the
+        // signaled amount. This amount is not vulnerable to being slashed but also does not accumulate rewards.
+
+        require(voterStakes[msg.sender].cumulativeStaked >= amount, "Bad request amount");
+
+        voterStakes[msg.sender].requestUnstake = amount;
```

Comment:
Since the user could technically reduce this value, is there any concern about users unstaking and then restaking those tokens before executing the withdrawal?

Reply:
you cant unstake until your liveness timer has passed. you must wait until the `unstakeCoolDown` time has passed from when you staked. if they change this value (i.e re-request to unstake) then this timer is reset.